File: backend/models/Schedule.py
# Schedule.py
# Schedule Object
from backend.models.Course import Course
ROUND = 2

class Schedule:
    def __init__(self, courses: list['Course'], score=None, weights=None):
        self.courses = courses
        self.score = score
        self.weights = weights

    def __repr__(self):
        result = f"score={self.score} | weights={self.weights}\n"
        result += "Subject   Credits CRN    Instructor                GPA   Days  Begin  End   Lab   Begin  End   Building\n"
        for course in self.courses:
            subject = course.subject
            course_credits = course.course_credits
            crn = course.crn
            instructor = course.instructor
            days = course.days
            start = course.start_time
            end = course.end_time
            lab_days = course.lab_days
            lab_start = course.lab_start_time
            lab_end = course.lab_end_time
            gpa = course.gpa
            room = course.room
            
            # Handling None values for lab related attributes
            lab_days_str = lab_start_str = lab_end_str = ""
            if lab_days is not None:
                lab_days_str = f"{lab_days:<5}"
            if lab_start is not None:
                lab_start_str = f"{lab_start:<4}"
            if lab_end is not None:
                lab_end_str = f"{lab_end:<4}"

            course_string = f"{subject:<9} {course_credits:7} {crn:6} {instructor:<25} {round(gpa, ROUND):<5} {days:5} {start:6} "\
                            f"{end:5} {lab_days_str:5} {lab_start_str:6} {lab_end_str:5} {room:6}\n"
            result += course_string
        
        return result

    def to_dict(self):
        return {
            "courses": [course.to_dict() for course in self.courses],
            "score": self.score,
            "weights": self.weights
        }
    
    def _weigh_gpa(self):
        # Filter and find gpa of schedule
        total_gpa = 0
        num_courses = 0
        for course in self.courses:
            if course.gpa:
                total_gpa += course.gpa
                num_courses += 1
        
        if num_courses == 0:
            return None
        
        # Calculate the average GPA, normalize it to a score between 0 and 1
        average_gpa = total_gpa / num_courses
        gpa_score = average_gpa / 4.0   # Weigh based on 4.0 scale     
        return round(gpa_score, ROUND)
    
    def _weigh_start(self, start_time):
        # Filter and find start time of schedule
        if not start_time:
            return None

        if start_time < 480 or start_time >= 780:  # Before 08:00 or after/equal to 13:00
            return 0.0
        elif 480 <= start_time < 600: # 0800 to 1000
            start_score = (start_time - 480) / 120.0
            return round(start_score, ROUND)
        elif 600 <= start_time < 660: # 1000 to 1100
            return 1.0
        else:
            start_score = (780 - start_time) / 120.0 # 1100 to 1300
            return round(start_score, ROUND) 
        
    def _weigh_end(self, end_time):
        # Filter and find end time of schedule
        if not end_time:
            return None
        
        if end_time <= 840: # 1400 in mins from midnight
            return 1.0
        elif end_time > 960: # 1600 in mins from midnight
            return 0.0
        else:
            end_score = (960 - end_time) / 120.0 # Between 1400 and 1600
            return round(end_score, ROUND)
        
    def _weigh_gaps(self, start_time, end_time):
        if not end_time or not start_time:
            return None

        day_time_mins = end_time - start_time
        
        course_time_mins = 0
        for course in self.courses:
            course_start = to_mins(course.start_time)
            course_end = to_mins(course.end_time)
            course_time_mins += (course_end - course_start)
        
        gap_time = day_time_mins - course_time_mins

        num_courses = len(self.courses)
        best_gaps_time = 10 * (num_courses - 1)

        if gap_time <= best_gaps_time:
            return 1.0
        elif gap_time > (120 + best_gaps_time):
            return 0.0
        else:
            # y=mx+b from best_gaps_time (1.0) to 120+best_gaps_time (0.0)
            gap_score = ((-1 / 120) * (gap_time - best_gaps_time)) + 1.0
            return round(gap_score, ROUND)

    def weigh_self(self):
        start_times = []
        for course in self.courses:
            if isinstance(course.start_time, str) and course.start_time.isdigit():
                start_times.append(int(course.start_time))
        start_time = to_mins(min(start_times))
    
        end_times = []
        for course in self.courses:
            if isinstance(course.end_time, str) and course.end_time.isdigit():
                end_times.append(int(course.end_time))
        end_time = to_mins(max(end_times))
        
        gpa_score = self._weigh_gpa()
        start_score = self._weigh_start(start_time)
        end_score = self._weigh_end(end_time)
        gap_score = self._weigh_gaps(start_time, end_time)

        self.weights = {
            "start": start_score,
            "end": end_score,
            "gap": gap_score,
            "gpa": gpa_score
        }

        scores = [value for value in self.weights.values() if value is not None]
        if scores:
            average = sum(scores) / len(scores)
            self.score = round(average, ROUND)
        else:
            self.score = None

# Returns minutes since midnight
def to_mins(time: str) -> int:
    time = int(time)
    return (time // 100) * 60 + (time % 100)

File: backend/internal/models/schedule.go
package models

import (
	"github.com/cwooper/schedule-optimizer/internal/utils"
)

var (
	dayBegin       = toMins(800)  // Converts the start of the day to mins
	dayEnd         = toMins(1700) // Converts the end of the day to mins
	startSlope     = 1.0 / float64(dayEnd-dayBegin)
	startIntercept = float64(dayBegin) / float64(dayEnd-dayBegin) * -1
	endSlope       = startSlope * -1
	endIntercept   = 1 + (startIntercept * -1)
)

// WeighFunc is a function type for weight calculation functions
type WeighFunc func(*Schedule) float64

// weightMap maps weight names to their corresponding weighing functions
var weightMap = map[string]WeighFunc{
	"GPA":   weighGPA,
	"GAP":   weighGap,
	"Start": weighStart,
	"End":   weighEnd,
}

// A weight for the schedule
type Weight struct {
	Name  string
	Value float64
}

// Holds a single displayable schedule object
type Schedule struct {
	Courses []Course // All courses in this schedule
	Weights []Weight // All current weights for this schedule
	Score   float64  // The average of the weights for an average score
}

// Creates a new schedule with the courses and initializes the weights
func NewSchedule(courses []Course) *Schedule {
	s := &Schedule{
		Courses: courses,
		Weights: make([]Weight, 0, len(weightMap)),
		Score:   0,
	}
	s.initializeWeights()
	s.weigh()
	return s
}

// initializeWeights initializes the Weights slice with all weights from weightMap
func (s *Schedule) initializeWeights() {
	for name := range weightMap {
		s.Weights = append(s.Weights, Weight{Name: name, Value: 0})
	}
}

// Weighs this schedule by evaluating and storing the internal weights
func (s *Schedule) weigh() {
	var totalWeight float64
	foundWeights := 0
	for i := range s.Weights {
		weighFunc, exists := weightMap[s.Weights[i].Name]
		if exists {
			s.Weights[i].Value = weighFunc(s)
			if s.Weights[i].Value != 0.0 {
				totalWeight += s.Weights[i].Value
				foundWeights += 1
			}
		} // weights already initialized to 0
	}
	s.Score = utils.Round(totalWeight / float64(foundWeights))
}

// ----------------------- Weighing Logic -----------------------

// TODO untested
func weighGPA(s *Schedule) float64 {
	totalGPA := 0.0
	numCourses := 0
	for _, course := range s.Courses {
		if course.GPA > 0 {
			totalGPA += float64(course.GPA)
			numCourses++
		}
	}

	if numCourses == 0 {
		return 0
	}

	averageGPA := totalGPA / float64(numCourses)
	gpaScore := averageGPA / 4.0 // Normalize weight based on 4.0 scale
	return utils.Round(gpaScore)
}

// TODO untested
func weighGap(s *Schedule) float64 {
	daySchedules := initializeDaySchedules()
	updateDaySchedules(s, daySchedules)
	averageGapScore := weighDaySchedules(daySchedules)
	return utils.Round(averageGapScore)
}

// TODO untested
func weighStart(s *Schedule) float64 {
	startTime, _ := findStartEndTimes(s) // Already converted to mins
	if startTime == -1 {
		return 0.0
	}

	if startTime <= dayBegin || startTime >= dayEnd { // Before 08:00 or After 17:00 is error
		return 0.0
	} else {
		return startSlope*float64(startTime) + startIntercept
	}
}

// TODO untested
func weighEnd(s *Schedule) float64 {
	_, endTime := findStartEndTimes(s) // Already converted to mins
	if endTime == -1 {
		return 0.0
	}

	if endTime <= dayBegin || endTime >= dayEnd { // Before 08:00 or After 17:00 is error
		return 0.0
	} else {
		return endSlope*float64(endTime) + endIntercept
	}
}

// Helper function to convert time string to minutes since midnight
func toMins(time int) int {
	return int(time/100)*60 + (time % 100)
}

// Helper function to find the earliest start time and latest end time of the schedule
func findStartEndTimes(s *Schedule) (int, int) {
	endTime := -1
	startTime := -1
	for _, course := range s.Courses {
		for _, session := range course.Sessions {
			if session.IsAsync || session.IsTimeTBD {
				continue // There are no times for async or timetbd sessions
			}
			sessionStart := toMins(session.StartTime)
			sessionEnd := toMins(session.EndTime)
			if startTime == -1 || sessionStart < startTime {
				startTime = sessionStart
			}
			if sessionEnd > endTime {
				endTime = sessionEnd
			}
		}
	}
	return startTime, endTime
}

// daySchedule holds the schedule information for a single day
type daySchedule struct {
	totalCourseTime int // Total time spent in courses
	firstCourseTime int // Start time of the first course
	lastCourseTime  int // End time of the last course
}

func initializeDaySchedules() map[rune]*daySchedule {
	daySchedules := make(map[rune]*daySchedule)
	for _, day := range []rune{'M', 'T', 'W', 'R', 'F'} {
		daySchedules[day] = &daySchedule{
			totalCourseTime: 0,
			firstCourseTime: -1,
			lastCourseTime:  -1,
		}
	}
	return daySchedules
}

// Won't add session to the schedules if it's async or tbd time
func updateDaySchedules(s *Schedule, daySchedules map[rune]*daySchedule) {
	for _, course := range s.Courses {
		for _, session := range course.Sessions {
			if !session.IsAsync && !session.IsTimeTBD {
				updateSessionSchedule(session, daySchedules)
			}
		}
	}
}

func updateSessionSchedule(session Session, daySchedules map[rune]*daySchedule) {
	sessionStartMins := toMins(session.StartTime)
	sessionEndMins := toMins(session.EndTime)
	sessionDuration := sessionEndMins - sessionStartMins

	for _, day := range session.Days {
		schedule := daySchedules[day]
		schedule.totalCourseTime += sessionDuration
		updateDayBoundaries(schedule, sessionStartMins, sessionEndMins)
	}
}

func updateDayBoundaries(schedule *daySchedule, startMins, endMins int) {
	if schedule.firstCourseTime == -1 || startMins < schedule.firstCourseTime {
		schedule.firstCourseTime = startMins
	}
	if schedule.lastCourseTime == -1 || endMins > schedule.lastCourseTime {
		schedule.lastCourseTime = endMins
	}
}

// Weighs the daySchedules based on their total time vs gap time
func weighDaySchedules(daySchedules map[rune]*daySchedule) float64 {
	totalGapScore := 0.0
	activeDays := 0

	for _, schedule := range daySchedules {
		if schedule.totalCourseTime > 0 {
			activeDays++
			totalDayTime := schedule.lastCourseTime - schedule.firstCourseTime
			gapTime := totalDayTime - schedule.totalCourseTime
			gapRatio := float64(gapTime) / float64(totalDayTime)
			totalGapScore += 1 - gapRatio // Higher score for less gap time
		}
	}

	if activeDays == 0 {
		return 0
	}

	return totalGapScore / float64(activeDays)
}
